### PR TITLE
Zolaification Proposal

### PR DESCRIPTION
**Pull request by XAMPPRocky**
### [Rendered](https://github.com/XAMPPRocky/web-presence-team/blob/patch-1/website-team/proposals/zola.md)

### CODE REVIEW

File: website-team/proposals/zola.md [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+- **Tera templating language —** Tera is Zola's backing templating language that is a much more fully featured templating language that allows us to write a lot of the logic that we have in Rust directly in the Tera templates themselves.
+- **Written In Rust —** Since Zola is written in Rust, it's much easier for Rust Web Presence members to contribute features to Zola that they need for the website. In fact some of that is currently in progress to add features currently missing from Zola that is present in our bespoke apps such as fluent support and multiple RSS feeds.
+- **Asset collocation —** Zola by default allows you to host static content from the same directory as the source markdown, this removes some of the absolute URLs that required on the blog because the static assets directory is seperate.
+
+
+## URL Scheme Changes
+
+While Zola does provide the functionality we need it does not map directly to our current URL scheme, and merging the three websites will require moving the blog and thanks to be available under the `www.rust-lang.org` host. The blog currently has the following URL scheme.
```
Comment (emilyalbini) [inline website-team/proposals/zola.md:21]: That's not true, we can have a single codebase but websites hosted on different domains. We simply need to generate the three domains in directories, and then serve each directory under its own domain name:

```
build/
├── blog
├── thanks
└── www
    └── assets
```

File: website-team/proposals/zola.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+**Current Scheme**
+
+    https://blog.rust-lang.org/{YYYY}/{MM}/{DD}/{title}.html
+    https://blog.rust-lang.org/inside-rust/{YYYY}/{MM}/{DD}/{title}.html
+
+This date based URL scheme is not supported in Zola and the Zola maintainers have said they are not interested in adding this date scheme. Additionally we have had issues with the date based URL scheme, as sometimes a blog post is merged with the wrong date, and then updating the link will break any links to the previous posts, such as on social media or in RSS readers. For these reasons there is motivation to also take the time to use merging onto `www.rust-lang.org` to also revisit this URL scheme and then make the switch once.
```
Comment (emilyalbini) [inline website-team/proposals/zola.md:28]: > Additionally we have had issues with the date based URL scheme, as sometimes a blog post is merged with the wrong date, and then updating the link will break any links to the previous posts, such as on social media or in RSS readers.

This is a problem that can be solved in the blog repository, by enabling bors and adding a check on CI that prevents merging if the blog post date is different than the current one.

File: website-team/proposals/zola.md [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+    https://www.rust-lang.org/blog/{topic}/{title}.html
+    https://www.rust-lang.org/blog/inside-rust/{team}/{title}.html
+    # Examples
+    https://www.rust-lang.org/blog/releases/1.44.0.html
+    https://www.rust-lang.org/blog/docs-rs/choose-your-build-target.html
+    https://www.rust-lang.org/blog/inside-rust/governance-wg/meetings/2020-04-05.html
+## Migrating to Zola
+
+Luckily the website being divided into three separate apps actually makes migrating to a single component easier, as we can by switching only a single component and slowly integrate the other two without requiring that maintenance work be stopped or blocked on the transition. The best candidate to start with is the Rust blog, the website currently has fluent localisation which is not currently supported by Zola (though it is being actively worked on), and "Thanks" is both pretty bespoke in terms of functionality and is the least popular of the three. Switching the blog allows us to test and iterate on Zola with a sizeable portion of our audience, while maintaining the current level of functionality.
```
Comment (emilyalbini) [inline website-team/proposals/zola.md:42]: I'd start with thanks instead, as it's the simplest and there isn't much impact if something breaks.

**Review by emilyalbini — COMMENTED** [review]
Thanks for proposing this! Reducing the maintenance overhead and removing the backend application for the website are huge improvements over the status quo. There are some things I'm not fully in agreement with.

This doesn't mention the fate of the blog.rust-lang.org repository: I'd prefer for the repository containing the actual posts to be separate from the website codebase, to allow giving teams write access to it without giving access to the whole website codebase. We can integrate it with the main repository by having the blog's CI send a webhook and trigger a rebuild.

I also think we should keep the current URL schema and domain names as much as possible, as we have good SEO on them and it's not worth losing it.

Finally, there are some parts of the website that are actually dynamic: both the "users" and the "sponsors" page are randomized each visit, to avoid highlighting specific companies to much. How does your plan handle this?

File: website-team/proposals/zola.md [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+There are a lot of different static site generators, so there's a good reason ask why pick one over another. Zola has quite a few unique features that make it particularly appealing to us.
+
+
+- **Independently maintained. —** We're not the sole or the biggest user of Zola, so there is plenty of community interest in continuing to use and add to Zola.
+- **Tera templating language —** Tera is Zola's backing templating language that is a much more fully featured templating language that allows us to write a lot of the logic that we have in Rust directly in the Tera templates themselves.
+- **Written In Rust —** Since Zola is written in Rust, it's much easier for Rust Web Presence members to contribute features to Zola that they need for the website. In fact some of that is currently in progress to add features currently missing from Zola that is present in our bespoke apps such as fluent support and multiple RSS feeds.
+- **Asset collocation —** Zola by default allows you to host static content from the same directory as the source markdown, this removes some of the absolute URLs that required on the blog because the static assets directory is seperate.
+
+
+## URL Scheme Changes
```
Comment (steveklabnik) [inline website-team/proposals/zola.md:19]: This section does not talk about SEO implications of changing the URL scheme, nor how to mitigate the impacts.

File: website-team/proposals/zola.md (hunk 1 above)
Comment (XAMPPRocky) [reply website-team/proposals/zola.md:21]: That won't work because there will be shared assets such as fonts and CSS that are located in subdirectories from the root of the build directory. Those assets will not be accessible through those domains without having to complicate the build process, and getting running it locally harder. With the goal of simplification I would not want building the website to be more complicated than running `zola serve` or `zola build`.

File: website-team/proposals/zola.md (hunk 2 above)
Comment (XAMPPRocky) [reply website-team/proposals/zola.md:28]: I feel like that is solving the problem in the wrong way. Adding bors is a lot of overhead for a website, it's better to just update master and build again it's faster and bors doesn't actually solve the problem of the article being on the wrong date.

Instead of putting in additional work to add checks to stop a problem it's better to never have the problem because it's not part of URL scheme, and thus can be updated without breaking changes.

**Comment by XAMPPRocky** [comment]
> This doesn't mention the fate of the blog.rust-lang.org repository: I'd prefer for the repository containing the actual posts to be separate from the website codebase, to allow giving teams write access to it without giving access to the whole website codebase. We can integrate it with the main repository by having the blog's CI send a webhook and trigger a rebuild.

That's a lot work, and as I mentioned in another comment would really complicate building for CI and locally in development. I want to keep this as simple to use and develop for as possible. The only thing you should have to run when working on the site is `zola serve`.

> Finally, there are some parts of the website that are actually dynamic: both the "users" and the "sponsors" page are randomized each visit, to avoid highlighting specific companies to much. How does your plan handle this?

This is what I see as one of the main benefits of switching to Zola is that it is quite powerful and you can easily create these one-off components that you usually have on a website. I don't want to get bogged down in the details of those specific components, but I can provide an example from my [initial Zola blog PR][zpr].

The blog has a metadata parameter called `team` which was in the format of `the Secure Code WG <https://www.rust-lang.org/governance/wgs/wg-secure-code>`. Which is currently a simple string split. However with Tera it was actually easier to provide the team's key to get the team information directly from the team repo.  This is also provides a build error if someone creates a post for a team that doesn't exist.

**Metadata**
```markdown
+++
author = "Jane Doe"
[extra]
team = "governance-wg"
+++
```

**Template**
```django
    <div class="publish-date-author">{{ page.date | date(format="%b. %d") }} &middot; {{ page.extra.author }}

    {% if page.extra.team %}

        {# Generates "on behalf of <team>" using team data. #}
        {% set teams = load_data(url=config.extra.team_url, format="json") %}
        {% set team = teams | get(key=page.extra.team) %}
        {% set url = "https://www.rust-lang.org/governance/" %}

        {% if team.subteam_of %}
            {% set url = url ~ "teams/" %}
        {% elif team.kind == "working_group" %}
            {% set url = url ~ "wgs/" %}
        {% else %}
            {% set url = url ~ "teams/" %}
        {% endif %}

        {% if team.subteam_of %}
            {% set url = url ~ team.subteam_of ~ "#" ~ team.website_data.page %}
        {% else %}
            {% set url = url ~ team.website_data.page %}
        {% endif %}

        on behalf of the <a href="{{ url }}">{{ team.website_data.name }}</a>.
    {% endif %}
    </div>
```

[zpr]: https://github.com/XAMPPRocky/blog.rust-lang.org/tree/zola


> I also think we should keep the current URL schema and domain names as much as possible, as we have good SEO on them and it's not worth losing it.

Of course I would also like this if it were possible, but I think the core message of this proposal is that it's not going to be possible. It's simply not going to be possible to continue to maintain three separate websites and codebases at our current size. The website needs a dramatic reduction in its scope to be sustainable.

I don't mean single you out as multiple members have mentioned URLs as part of *"good SEO"* , and this has really bothered me, as it is frankly ignorant of actual SEO best practices.

Aside from the fixed cost of making any kind of change at all, you're not going to be able to tell whether a specific URL scheme provides better SEO results than another. Web crawlers handle 301 redirects just fine.

If we look at Google's [Webmaster Guidelines][webg] for what best practices they recommend for improving your web page SEO, you can see there are a number of guidelines such as having a `sitemap` file, and supporting the `If-Modified-Since` header that we are not currently doing that we would get **for free** with this transition to Zola and a single static site. So to me it seems like it will only improve the SEO.

[webg]: https://support.google.com/webmasters/answer/35769

File: website-team/proposals/zola.md (hunk 1 above)
Comment (emilyalbini) [reply website-team/proposals/zola.md:21]: I have looked a bit about Zola now, and with a slightly different structure I think it's possible to achive what I'm proposing.



This way one can run `zola serve` or `zola build` on the site they want to work on, and all the content we would share anyway is in the top-level directories, symlinked into each site.

**Comment by emilyalbini** [comment]
> That's a lot work, and as I mentioned in another comment would really complicate building for CI and locally in development. I want to keep this as simple to use and develop for as possible. The only thing you should have to run when working on the site is zola serve.

Your approach would really complicate landing blog posts though: for the website we have manual deploys by updating the `deploy` branch, and a `try` branch. Either we get rid of that (deploying translations directly, which I don't think is something we want) or we'd have to also deploy the existing changes every time we land a new post.

> This is what I see as one of the main benefits of switching to Zola is that it is quite powerful and you can easily create these one-off components that you usually have on a website. I don't want to get bogged down in the details of those specific components, but I can provide an example from my initial Zola blog PR.

Sorry, I should've explained myself better: I was talking about having content that is randomly sorted every time a user visits a page, not every time the website is rebuilt.

> Of course I would also like this if it were possible, but I think the core message of this proposal is that it's not going to be possible. It's simply not going to be possible to continue to maintain three separate websites and codebases at our current size. The website needs a dramatic reduction in its scope to be sustainable.

Sure, but changing the whole URL scheme is not *required* to switch to a single codebase and a static generator. If that's actually not possible it means the generator is not up to the task, and maybe we should consider using a different one.

Of course there is no difference between two URL schemes over time, but we'd be making a change we're not required to do.

**Comment by XAMPPRocky** [comment]
> Your approach would really complicate landing blog posts though: for the website we have manual deploys by updating the deploy branch, and a try branch. Either we get rid of that (deploying translations directly, which I don't think is something we want) or we'd have to also deploy the existing changes every time we land a new post.

Well I initially left out the website's deployment as I thought it would be more appropriate as a separate discussion, but I do think the website's deployment story should be like forge or the blog. Having a separate branch for deployments is required when you have a open source rocket backend where it's an executable running somewhere.

However once we can have a static content deployed with GitHub Pages, I don't think there's much reason to keep a deploy branch. We don't get bug reports from people using staging, and keeping the production website up to date is extra busy work for the maintainers.

I would expect that we would continue to have the whitelist of languages that production ready. Besides that I think we want the website's content to be easy for contributors to change and add new content. The website's information today is inconsistent and quite a few places out of date.

> Sorry, I should've explained myself better: I was talking about having content that is randomly sorted every time a user visits a page, not every time the website is rebuilt.

Well it's a couple of lines of JavaScript to randomise a list of DOM elements. I'm not really sure why this relevant even if some things don't map one-to-one, I think it's better to replicate intent rather behaviour. For example I don't see why it wouldn't be acceptable if the list was generated at build time and rebuilt at least once a day (which is very easy to do with GitHub Actions)

> Sure, but changing the whole URL scheme is not required to switch to a single codebase and a static generator. If that's actually not possible it means the generator is not up to the task, and maybe we should consider using a different one.
> Of course there is no difference between two URL schemes over time, but we'd be making a change we're not required to do.

I don't think we should let perfect be the enemy of good. While it's possible for a static site generator to exist that could meet all of these requirements. We have to look at what's currently available and what's feasible, particularly in the near term.

I'm not going to say changing the URL scheme it's an easy change to make, but it is a mostly *one-time* cost in setting up and ensuring the redirects are working versus structuring our entire code and CI around these vanity URLs.

**Comment by emilyalbini** [comment]
> Well I initially left out the website's deployment as I thought it would be more appropriate as a separate discussion, but I do think the website's deployment story should be like forge or the blog. Having a separate branch for deployments is required when you have a open source rocket backend where it's an executable running somewhere.
>
> However once we can have a static content deployed with GitHub Pages, I don't think there's much reason to keep a deploy branch. We don't get bug reports from people using staging, and keeping the production website up to date is extra busy work for the maintainers.
>
> I would expect that we would continue to have the whitelist of languages that production ready. Besides that I think we want the website's content to be easy for contributors to change and add new content. The website's information today is inconsistent and quite a few places out of date.

Makes sense. Staging could still be useful to preview how a design change would look though.

> Well it's a couple of lines of JavaScript to randomise a list of DOM elements. I'm not really sure why this relevant even if some things don't map one-to-one, I think it's better to replicate intent rather behaviour. For example I don't see why it wouldn't be acceptable if the list was generated at build time and rebuilt at least once a day (which is very easy to do with GitHub Actions)

I think having those lists randomized every time a user visits the page is really important, and that's a disadvantage of statically generated websites we need to be aware of. I'm not advocating for staying with the dynamic application, we just need to have an acceptable solution ready, and the JS one is probably the best one yet.

> I don't think we should let perfect be the enemy of good. While it's possible for a static site generator to exist that could meet all of these requirements. We have to look at what's currently available and what's feasible, particularly in the near term.
>
> I'm not going to say changing the URL scheme it's an easy change to make, but it is a mostly one-time cost in setting up and ensuring the redirects are working versus structuring our entire code and CI around these vanity URLs.

The Zola maintainer's stance worries me a bit though: a static generator shouldn't add opinionated *restrictions* to a website, especially when you're willing to implement that feature yourself.

We can already avoid merging everything in www.rust-lang.org with the multiple Zola projects trick, and it'd be really sad if we'd have to change our URL scheme just for the blog because the maintainer refuses to implement support.

**Comment by XAMPPRocky** [comment]
> Makes sense. Staging could still be useful to preview how a design change would look though.

Well a person could enable GitHub Pages on their fork and change the `BASE_URL` in the config and it would be available rendered on GitHub Pages.  I was thinking as a wishlist that it could be automated by a bot that pushes a build to a GitHub Pages repo, but doing it manually as needed seems like a reasonable compromise for in the mean time.

> The Zola maintainer's stance worries me a bit though: a static generator shouldn't add opinionated restrictions to a website, especially when you're willing to implement that feature yourself.

> We can already avoid merging everything in www.rust-lang.org with the multiple Zola projects trick, and it'd be really sad if we'd have to change our URL scheme just for the blog because the maintainer refuses to implement support.

Well I disagree with your opinion on static generators, the moment you choose or not choose a CSS pre-processor you've made a very opinionated decision on how sites that use your tool should be structured, so there's no avoiding having an opinion on a websites structure. And I would add that implementing a feature isn't the hard part, it's the continued maintenance. Which is why I'm very resistant to add any kind of hack or trick especially at the start of a rewrite, as I know for a fact that there will be more work required to keep it working in the future, and I don't want dedicate what limited time I have to fixing the hack if I never have to put it in the first place.

Either way though I don't feel this is related to Zola. Rust's current website architecture is the result of historical accident not deliberate design, it's not surprising that it doesn't fit. All three of these sites started as separate projects which is why they are on separate sub-domains.  If we want them to present them as a single project (which was my impression), then it should be structured as a single website, not as three. And if we want to keep them as three we shouldn't be trying to merge them, and we should pull out the website styling on the other projects and present them as separate websites.

I can't stress enough how much maintaining three separate base URLs for a single website is a net negative change in every way except that it preserves the status quo. As I've mentioned I have huge maintenance and contributor concerns around implementing a bespoke hack, but having different URLs will provide worst performance, as a user's browser can't cache shared resources for free when going between the blog, thanks, and the main website. Which is materially meaningful on mobile and low-bandwidth regions.

**Comment by emilyalbini** [comment]
> Well a person could enable GitHub Pages on their fork and change the BASE_URL in the config and it would be available rendered on GitHub Pages. I was thinking as a wishlist that it could be automated by a bot that pushes a build to a GitHub Pages repo, but doing it manually as needed seems like a reasonable compromise for in the mean time.

Makes sense.

> I can't stress enough how much maintaining three separate base URLs for a single website is a net negative change in every way except that it preserves the status quo. As I've mentioned I have huge maintenance and contributor concerns around implementing a bespoke hack, but having different URLs will provide worst performance, as a user's browser can't cache shared resources for free when going between the blog, thanks, and the main website. Which is materially meaningful on mobile and low-bandwidth regions.

While I'd prefer not to change the URL scheme if possible, I understand that I'm usually not happy about changing existing stuff when we're not required to so. If all the other members of the website team are happy with changing the URL scheme I won't be the one blocking it.

@skade @Manishearth what's your opinion on this?

**Comment by Manishearth** [comment]
I haven't read through everything and will soon but I'm strongly against merging the repos, the blog is high traffic for markdown content (which I don't need to review, blog posts are reviewed by individual teams), especially after inside-rust, whereas the website is more mixed traffic for both content and design tweaks. It would be much harder to properly maintain the website if it were merged as a repo. The blog has a very different cadence and merging the projects doesn't make sense to me.

I'm open to solutions with more repos that let us share the styling.

**Comment by XAMPPRocky** [comment]
> I haven't read through everything and will soon but I'm strongly against merging the repos, the blog is high traffic for markdown content (which I don't need to review, blog posts are reviewed by individual teams), especially after inside-rust, whereas the website is more mixed traffic for both content and design tweaks. It would be much harder to properly maintain the website if it were merged as a repo. The blog has a very different cadence and merging the projects doesn't make sense to me.
>
> I'm open to solutions with more repos that let us share the styling.

Well to be clear while the teams do often review their own posts on the blog they don't have merge rights and there's still a lot of the review and procedural work being done by myself, @pietroalbini, and @nikomatsakis. For me it is a lot harder to maintain now than it would be as a single repository as I can't share a lot of stuff like CI, labels, or issue templates.

It should also be noted that it's not just styling that needs to be shared, it's also content such as in the header and footer that need to be shared. It is my belief that the website is low traffic in contributions because it is hard for its target audience to contribute to it.

This was true with forge, where last year it was based in Jekyll and a flat list of articles, it was a pain to maintain and a lot of the information was out of date and focused entirely on infra and release. However with the rewrite to mdbook and restructuring the content to be grouped by team, it's now full of content for nearly every team.

To me the target contributor audience for the website is frontend devs, UI/UX designers, and content writers who are interested in Rust. With them in mind I don't think you should have to write Rust or mess with git just to see how your content looks, or add new design components.

**Comment by Manishearth** [comment]
> Having a separate branch for deployments is required when you have a open source rocket backend where it's an executable running somewhere.

I think it's still important for styling changes to have an extra wait time; we have had issues in the past where we've landed things

An alternate solution for that is to have automation that produces rendered pages that can be visited per-PR. Not sure how hard this is, I've seen this kind of thing before for singlepage renders but not for entire websites.
 
Or, we can have a policy of running styling changes through a staging branch.

> It should also be noted that it's not just styling that needs to be shared, it's also content such as in the header and footer that need to be shared.

Right, but all of this can still be done by submoduling. Alternatively the _content_ can be a submodule, though this makes deploying annoying. I really don't want the blog content to overlap with the website/design/layout when it comes to the repository. 


> Well it's a couple of lines of JavaScript to randomise a list of DOM elements.

Yeah I think we should just do this

-----

Another thing I didn't realize is that the base domain for the blog is changing as well. While I'm all for adding more redirects, I'm very wary of changing canonical URLs.

This isn't just a matter of status quo for status quo's sake, I prefer a separate "blog" domain, it comes up differently on link aggregator sites/Google/etc.

Furthermore, it's a bit of a matter of community memory as well, folks expect things to be on blog.rust-lang.org (I care less about thanks.rlo).

Duplicate downloads on slow connections is an issue, but overall the real problem with slow connections is making sure you don't have _too much_ JS/CSS (especially synchronous JS loads), not as much as "oh oops I downloaded this twice". Our files at the moment are not of a size where these things become a problem. I've spent significant amounts of time on browsers with slow connections, our website is nowhere near a point where anything we do there will matter. Slow connections aren't ever _that_ slow.

> It's simply not going to be possible to continue to maintain three separate websites and codebases at our current size

It's not clear to me that _merging domains_ is the only way to achieve this goal. You can have split domains and repos but still share most of the layout and styling code. I can easily imagine a root repo that contains all of the layout/styling stuff (plus some FTL files), and it's submoduled inside a repo that contains the content and the rest of the FTL files, and your build step makes it work.

**Comment by XAMPPRocky** [comment]
> I think it's still important for styling changes to have an extra wait time; we have had issues in the past where we've landed things

Blog posts are released with typos and the production version does have issues despite the process. it’s easier to fix bugs as they come up on a static website since the total time from CI to production is less than two minutes. It’s much less work than managing a release cycle.

> Furthermore, it's a bit of a matter of community memory as well, folks expect things to be on blog.rust-lang.org (I care less about thanks.rlo).

Yes and it’s not going away, blog.rust-lang.org will still bring them to the blog, just through a redirect.

> It's not clear to me that merging domains is the only way to achieve this goal. You can have split domains and repos but still share most of the layout and styling code. I can easily imagine a root repo that contains all of the layout/styling stuff (plus some FTL files), and it's submoduled inside a repo that contains the content and the rest of the FTL files, and your build step makes it work.

I’m not trying to put forward this proposal as the only way. I’m sure there are plenty of ways we can imagine and design to meets all of these requirements. However my main point behind this proposal is simplification and I feel like this is getting away from that.

Part of what makes this change so compelling for me is that it provides so many benefits with little to no real user impact and it has such a small and clear scope. I have [already rewritten][zpr] the blog in Zola, I started out just trying to get one post working and it ended up taking only half a day to rewrite the whole blog. I’m very confident we could have a working version of the main website in a week (I already have a local Zola fork with fluent support), and most of that week would be spent on integrating the frontend rather than replacing the backend.

“Thanks” will actually probably take the longest since it’s very bespoke, but even then I think it can be done easily by just changing the existing codebase to also publish JSON on GitHub which can then be used at build time by the Zola site.

Another benefit of collapsing down to a single domain is we could handle this transition iteratively in production by having one of the base URLs such as the blog point to the Zola instance while keeping the others at their current codebases until it’s ready.

These are benefits you mostly get for free with the single domain approach since it is the most common approach on the web. With a multi-domain multi-repo setup you would have to engineer all the work yourself to achieve the same effect.

[zpr]: https://github.com/XAMPPRocky/blog.rust-lang.org/tree/zola

**Comment by Manishearth** [comment]
> Blog posts are released with typos and the production version does have issues despite the process

Blog posts do not have the same prominence as the website itself. It's fine if the blog is down for a bit, it's not fine if the website is. In the past we've had bugs that completely broke things which is why we instituted the deploy step.

I suspect there will be less chances for failure for a static site, however. It might be good to merge most things directly to master, but select things to a staging branch.

> However my main point behind this proposal is simplification and I feel like this is getting away from that.

I'm not a fan of making qualitative changes for implementation ease.

The multi-repo approach has a lot of benefits: It allows us to keep separate domains, it allows the review group to be different for the website vs the blog, and it makes it easier to ensure that no website content is untranslated (blog content will not be).

> With a multi-domain multi-repo setup you would have to engineer all the work yourself to achieve the same effect.

I don't see how this is true. The multi repo setup is just splitting the existing setup into multiple repos that use submodules so that the layout/config can mostly be shared. There may need to be an additional build step of copying files and changing base URLs,  but it's not major.

**Comment by XAMPPRocky** [comment]
> Blog posts do not have the same prominence as the website itself. It's fine if the blog is down for a bit, it's not fine if the website is. In the past we've had bugs that completely broke things which is why we instituted the deploy step.
> 
> I suspect there will be less chances for failure for a static site, however. It might be good to merge most things directly to master, but select things to a staging branch.

Can you provide a specific example of a failure that would be so bad on a static site that would merit a staging branch? If for any reason the website doesn't build, it also doesn't deploy so the previous version of the website will still be available. Keeping the website up is completely moved to our third party providers (GitHub Pages, Cloudfront, Gandi).

> The multi-repo approach has a lot of benefits: It allows us to keep separate domains, it allows the review group to be different for the website vs the blog

There's no reason we can't have different review groups in a mono repo, we do this already today. As far as I know we have access to the `CODEOWNERS` feature, so we can very easily divide a mono repo among different reviewers.

>  and it makes it easier to ensure that no website content is untranslated (blog content will not be).

Well there seems to be some confusion here because multiple people have told me that they want the blog's content to be translatable, and you're the first person who has said otherwise (cc @skade and @vertexclique).

> I don't see how this is true. The multi repo setup is just splitting the existing setup into multiple repos that use submodules so that the layout/config can mostly be shared. 

If it is just that, then feel free to create a proof of concept. I don't think I can simply take it on faith that an entirely unproven and bespoke approach will be a net positive or as simple as you make out it to be.

> There may need to be an additional build step of copying files and changing base URLs, but it's not major.

While you might not view it as major it is more than nothing and that is meaningful. We all use Cargo, I don't think I need to explain how much easier it is to contribute to Rust projects because in nearly every project all you need to do is run `cargo build && cargo test` with zero setup.

**Comment by Manishearth** [comment]
> There's no reason we can't have different review groups in a mono repo, we do this already today. As far as I know we have access to the `CODEOWNERS` feature, so we can very easily divide a mono repo among different reviewers.

For watching the repo I mean



> Well there seems to be some confusion here because multiple people have told me that they want the blog's content to be translatable, and you're the first person who has said otherwise (cc @skade and @vertexclique).

It shouldn't use fluent, either way, is what I mean, it will be translated as separate markdown files that are stored alongside the others in the appropriate directory structure.

> While you might not view it as major it is more than nothing and that is meaningful

I mean, for local dev the base URL doesn't matter at all (you're probably going to have it default to localhost anyway!), this is only for deploy.

**Comment by Manishearth** [comment]
> Can you provide a specific example of a failure that would be so bad on a static site that would merit a staging branch? 

Header/footer text breaking due to a change in the translation strings, for one.

Anyway this is mostly a "we can handle it when we have changes of this form". I suspect most will not be.

**Comment by XAMPPRocky** [comment]
> It shouldn't use fluent, either way, is what I mean, it will be translated as separate markdown files that are stored alongside the others in the appropriate directory structure.

Sure, Zola already has this kind of localisation system [already available today][zm] that will work in addition to Fluent.

[zm]: https://www.getzola.org/documentation/content/multilingual/

> I mean, for local dev the base URL doesn't matter at all (you're probably going to have it default to localhost anyway!), this is only for deploy.

Well no actually Zola is bit more clever than this, and this is missing my point. For local development you run `zola serve` which uses `localhost` as the base URL, and when you use `zola build` it uses the config's base URL so you should never have to change it for local development.

The problem I'm talking about is that since they are three websites in order to fully test it you have to three times the work for virtually no benefit. Since they are three different websites and separate Zola instances we can't use relative URLs or Zola's framework to get URLs to pages that are on a website different from the one they're currently on (www linking to blog, vice versa, etc).

So in order to test changes locally each website needs two extra config variables of the base URL of the other two websites in order to properly able to have URLs go to local versions of that website. That's six changes across three different files for no reason just to get started that we wouldn't need to do if it was a single website.

Not even to get into the fact that we don't have a easy way to pass state with static websites so people will lose their locale for example if they go from www to thanks and back, a problem we have right now.

Overall merging into a single website brings a lot of benefits to the user and developer experience, reduces complexity, and eases maintenance with no real breaking changes. Where as with maintaining three separate websites we have to do a lot of extra work just to be consistent and maintain the appearance of a single website. I still hold the position that we should either merge three existing websites and have a single website with a single design, or make the designs of the blog and thanks more distinct and paired down so that they don't require extra work to be consistent with www.

**Comment by Manishearth** [comment]
> Not even to get into the fact that we don't have a easy way to pass state with static websites so people will lose their locale for example if they go from www to thanks and back, a problem we have right now.

They .... won't, you just need all URLs to use `domain/{{lang}}/foo`. This is what we do right now anyway.

> So in order to test changes locally each website needs two extra config variables of the base URL of the other two websites in order to properly able to have URLs go to local versions of that website.

I don't see some extra config vars to be a major issue. I think it's fine that when testing locally you test one site at a time.

--------

Also, to be clear:

I'm totally okay with it ending up as a single Zola site. I don't think that's incompatible with anything I'm suggesting.

What I am suggesting is:

 - the blog content be maintained in a separate repo (this can be done as a submodule of markdown files, with CI autoupdating it).
 - the production blog use blog.rust-lang.org. It can be generated at `/blog`, we just use `{{blogbaseurl}}` or something for links between blog pages and point nginx at the folder.

**Comment by XAMPPRocky** [comment]
> They .... won't, you just need all URLs to use domain/{{lang}}/foo. This is what we do right now anyway.

This isn't what we do today, you can very easily replicate this.

1. Go to https://www.rust-lang.org/es/
2. Click on "Blog"
3. Click on "Rust" you'll be back in English.

Either way the complication I meant comes from either having separate fluent projects which might not always sync up so you might have a locale available in www but in blog and still have the same problem. Or having an additional shared locales submodule between all of the websites which is its own can of worms.

> the blog content be maintained in a separate repo (this can be done as a submodule of markdown files, with CI autoupdating it).

Well I looked into what this could look like and couldn't find a good solution. As blog content isn't just markdown, with Zola we will be able to have extra content like shortcodes. The front matter in the markdown also needs to be validated by Zola from the main website.

I don't know if agree that the adding the blog will add that much more traffic for watchers. On average there's one to two posts a week as well as occasional typo fix PRs. Most the current code traffic is updating dependencies and keeping it in sync with www that would disappear once integrated.

> the production blog use blog.rust-lang.org. It can be generated at /blog, we just use {{blogbaseurl}} or something for links between blog pages and point nginx at the folder.

Would the solution as described in the stack Overflow linked below be satisfactory where it essentially happens entirely on nginx?

[_"rewrite - Nginx URL masking to a different domain - Stack Overflow"_](https://stackoverflow.com/questions/27645018/nginx-url-masking-to-a-different-domain)

**Comment by Manishearth** [comment]
> This isn't what we do today, you can very easily replicate this.

This is what we do today for "state" within www, I mean. We [use a baseurl](https://github.com/rust-lang/www.rust-lang.org/blob/d8869aeaa2a54be6ffbd678ebe03787712270bcc/templates/components/footer.hbs#L22) that's defined to contain the language code.

>      Either way the complication I meant comes from either having separate fluent projects which might not always sync up so you might have a locale available in www but in blog and still have the same problem

Seems fine to me, the footer/etc would be localized and the blog content would fall back. Already the case for any new content that gets added to www.

> I don't know if agree that the adding the blog will add that much more traffic for watchers. 

That's fair. Niko had an alternate suggestion to use highfive's watcher ping functionality so that core team members (and other interested parties) can "watch" the blog.rlo folder without being bombarded with everything.

> Would the solution as described in the stack Overflow linked below be satisfactory where it essentially happens entirely on nginx?

In part, you also need all the `/blog/foo` links to say `blog.rust-lang.org/foo` instead. I believe this can be done in Zola by creating a separate `blogbaseurl` variable that's blog.rlo in production and localhost/blog in dev, and using it for all links to blog content.

**Comment by skade** [comment]
I'd like to make an effort to streamline the discussion a bit by separating decision subjects while writing down my thoughts.

## Zolaification in general

I think moving away from rocket is useful. Many of the features in the webapp don't need rocket and some are even a little painful to build in it. It was a very good choice when it was made, especially from the perspective of maintenance, but I think simpler approaches are now feasible.

I'm slightly positive towards zola.

## Domain: blog.rust-lang.org vs. www.rust-lang.org

The historical reason for this split is IMHO mainly historical: it's easier to deploy 2 different stacks on 2 domains using stock software and without relying on URL rewriting.

From a web property perspective, it's nowadays recommended to keep you blog under the main domain. This is a recommendation, not a must, though. I don't see a _must_ argument for unifying both, but also none against it.

## Separate blog repository

I think having a separate blog repository is useful. This is mainly for editors sake. It's important to make sure that editing people are focused on task at hand, having half another website carried around along with it is annoying.

I see the suggested middleground as the best: move the blog design into the website, but keep the blog content as a separate repository that is checked out during building the main website (or as a submodule).

## Blog i18n

I don't remember blog i18n ever being discussed as a currently desirable feature, except as an idea in passing. I do know that I have mentions to be part of the ruby-lang german team for a while. The Ruby blog can be translated freely by the teams, but doesn't require them to. This means that great care needs to be taken in the listing views: if an article isn't translated, the original will be presented instead.

All in all, if that ends up being a cheap feature, offering that makes sense. 

## Legacy URL stability

I have a strong stance on this: We must make sure all old URLs still work, either as a redirect or by still existing. The Rust website is linked in the media, we don't want to break those links. Redirects are fine, IMHO.

## URL format

It's common nowadays to not include the date in blog post URLs. That has multiple reasons:

* It doesn't make still-relevant posts look dated, just by looking at the URL
* The date in the URL has no value and doesn't differentiate the posts
* The date marker is prone to move, as illustrated by @XAMPPRocky, and adds an additional step during publishing.

I wouldn't see it as a blocker if it went away or were changed, given that we can establish URL stability.

**Comment by XAMPPRocky** [comment]
Well I think I want to write a new draft of the proposal that includes the discussion so far, and make a new PR for it when it's ready. Thanks everyone for the feedback!
